File: backend/internal/api/stations/wallpaper.go
package stations

import (
	"context"
	"log/slog"
	"net/http"
	"strconv"
	"sync"

	"github.com/LuukBlankenstijn/loom/backend/internal/domain"
)

func (s *stationsServer) WallpaperHandler(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	ip := getRealIP(r)

	var (
		contest *domain.Contest
		team    *domain.Team
		err     error
		wg      sync.WaitGroup
	)

	wg.Add(2)

	// Fetch Contest and Team in parallel to minimize external API latency
	go func() {
		defer wg.Done()
		contest, err = s.contestRepo.GetNextContest(ctx)
	}()

	go func() {
		defer wg.Done()
		team, _ = s.teamRepo.GetByIp(ctx, ip)
	}()

	wg.Wait()

	if err != nil {
		slog.Error("failed to get wallpaper, contest fetch failed", slog.Any("err", err))
		http.Error(w, "failed to get contest context", http.StatusNotFound)
		return
	}

	// Ensure the memory cache is in sync with the current contest
	if err := s.updateCache(ctx, contest.Id); err != nil {
		slog.Error("cache refresh failed", "err", err)
		http.Error(w, "wallpaper not set", http.StatusServiceUnavailable)
		return
	}

	s.wallpaperCache.mu.RLock()
	cached := s.wallpaperCache.wallpaper
	s.wallpaperCache.mu.RUnlock()

	if cached == nil {
		http.Error(w, "wallpaper not initialized", http.StatusServiceUnavailable)
		return
	}

	// Set dynamic headers based on the fetched team
	if team != nil {
		w.Header().Set("X-Wallpaper-Text", team.Name)
		w.Header().Set("X-Wallpaper-Text-Color", cached.TextColor)
	}

	w.Header().Set("Content-Type", cached.MimeType)
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Content-Length", strconv.Itoa(len(cached.Data)))

	// Serve the cached blob directly from memory
	w.Write(cached.Data)
}

func (s *stationsServer) updateCache(ctx context.Context, contestID string) error {
	// Fetch only the timestamp to check for freshness
	dbUpdatedAt, err := s.wallpaperRepo.GetLastUpdated(ctx, contestID)
	if err != nil {
		return err
	}

	// Fast path: check if cache is already current
	s.wallpaperCache.mu.RLock()
	isCurrent := s.wallpaperCache.wallpaper != nil &&
		s.wallpaperCache.wallpaper.UpdatedAt.Equal(dbUpdatedAt)
	s.wallpaperCache.mu.RUnlock()

	if isCurrent {
		return nil
	}

	s.wallpaperCache.mu.Lock()
	defer s.wallpaperCache.mu.Unlock()

	// Double-check inside lock to handle concurrent requests
	if s.wallpaperCache.wallpaper != nil &&
		!s.wallpaperCache.wallpaper.UpdatedAt.Before(dbUpdatedAt) {
		return nil
	}

	// Fetch the full blob only when necessary
	wallpaper, err := s.wallpaperRepo.GetWallpaper(ctx, contestID)
	if err != nil {
		return err
	}

	s.wallpaperCache.wallpaper = wallpaper
	return nil
}

func getRealIP(r *http.Request) string {
	IPAddress := r.Header.Get("X-Real-IP")
	if IPAddress == "" {
		IPAddress = r.Header.Get("X-Forwarded-For")
	}
	if IPAddress == "" {
		IPAddress = r.RemoteAddr
	}
	return IPAddress
}

File: backend/src/api/wallpaper.rs
use std::sync::Arc;

use axum::extract::{Query, State};
use axum::http::{HeaderMap, StatusCode, header};
use axum::response::IntoResponse;
use chrono::{DateTime, Utc};
use serde::Deserialize;
use tokio::sync::RwLock;

use tracing::error;

use crate::domain::{ContestRepository, TeamRepository, WallpaperRepository};

struct CachedWallpaper {
    data: Vec<u8>,
    mime_type: String,
    text_color: String,
    updated_at: DateTime<Utc>,
}

pub struct WallpaperState {
    contest_repo: Arc<dyn ContestRepository>,
    team_repo: Arc<dyn TeamRepository>,
    wallpaper_repo: Arc<dyn WallpaperRepository>,
    cache: RwLock<Option<CachedWallpaper>>,
}

#[derive(Deserialize)]
pub struct WallpaperParams {
    pub ip: Option<String>,
}

impl WallpaperState {
    pub fn new(
        contest_repo: Arc<dyn ContestRepository>,
        team_repo: Arc<dyn TeamRepository>,
        wallpaper_repo: Arc<dyn WallpaperRepository>,
    ) -> Self {
        Self {
            contest_repo,
            team_repo,
            wallpaper_repo,
            cache: RwLock::new(None),
        }
    }

    async fn update_cache(&self, contest_id: &str) -> Result<(), StatusCode> {
        let db_updated_at = self
            .wallpaper_repo
            .get_last_updated(contest_id)
            .await
            .map_err(|e| {
                error!(%e, "failed to get wallpaper last_updated");
                StatusCode::SERVICE_UNAVAILABLE
            })?;

        // Fast path: check if cache is current
        {
            let cache = self.cache.read().await;
            if let Some(cached) = &*cache
                && Some(cached.updated_at) == db_updated_at
            {
                return Ok(());
            }
        }

        // Slow path: fetch full blob under write lock
        let mut cache = self.cache.write().await;

        // Double-check after acquiring write lock
        if let Some(cached) = &*cache
            && Some(cached.updated_at) == db_updated_at
        {
            return Ok(());
        }

        let wp = self
            .wallpaper_repo
            .get_wallpaper(contest_id)
            .await
            .map_err(|e| {
                error!(%e, "failed to get wallpaper");
                StatusCode::SERVICE_UNAVAILABLE
            })?;

        *cache = wp.map(|w| CachedWallpaper {
            data: w.data,
            mime_type: w.mime_type,
            text_color: w.text_color,
            updated_at: w.updated_at,
        });

        Ok(())
    }
}

pub async fn wallpaper_handler(
    Query(params): Query<WallpaperParams>,
    State(state): State<Arc<WallpaperState>>,
) -> Result<impl IntoResponse, StatusCode> {
    let ip = params.ip;

    let contest = state
        .contest_repo
        .get_next_contest()
        .await
        .map_err(|e| {
            error!(%e, "failed to get next contest");
            StatusCode::NOT_FOUND
        })?
        .ok_or(StatusCode::NOT_FOUND)?;

    state.update_cache(&contest.id).await?;

    let cache = state.cache.read().await;
    let cached = cache.as_ref().ok_or(StatusCode::SERVICE_UNAVAILABLE)?;

    let mut response_headers = HeaderMap::new();
    if let Ok(v) = cached.mime_type.parse() {
        response_headers.insert(header::CONTENT_TYPE, v);
    }
    response_headers.insert(header::CACHE_CONTROL, "no-cache".parse().unwrap());

    create_headers(state.clone(), ip, &mut response_headers).await?;

    Ok((response_headers, cached.data.clone()))
}

async fn create_headers(
    state: Arc<WallpaperState>,
    ip: Option<String>,
    headers: &mut HeaderMap,
) -> Result<(), StatusCode> {
    match ip {
        None => Ok(()),
        Some(ip) => {
            let team = state
                .team_repo
                .get_by_ip(&ip)
                .await
                .inspect_err(|e| error!(%e, %ip, "failed to get team by ip"))
                .ok()
                .flatten();

            if let Some(team) = team {
                if let Ok(v) = team.name.parse() {
                    headers.insert("X-Wallpaper-Text", v);
                }
                let cache = state.cache.read().await;
                let cached = cache.as_ref().ok_or(StatusCode::SERVICE_UNAVAILABLE)?;
                if let Ok(v) = cached.text_color.parse() {
                    headers.insert("X-Wallpaper-Text-Color", v);
                }
            }
            Ok(())
        }
    }
}
